### DISCUSSION TITLE
What does the API "getSystemConsumedEnergy" get?

### DISCUSSION THREAD

**Discussion by baihe-liu** [opcm/pcm#904] [Q&A]
I'm trying to use pcm API to get energy consumption, my platform is Tiger Lake.
Could you please tell me what does system consumed energy means in the function "getSystemConsumedEnergy"?

```
template <class CounterStateType>
uint64 getSystemConsumedEnergy(const CounterStateType& before, const CounterStateType& after)
{
    return after.systemEnergyStatus - before.systemEnergyStatus;
}
```
I found SystemCounterState.systemEnergyStatus will only be updated by reading from MSR in function `PCM::getALLCounterStates`
And I got different result using SystemCounterState from 'getSystemCounterState' and 'getALLCounterStates'
```
================== Energy Consumption by getSystemCounterState  ==================  
[Energy_Monitor] getConsumedJoules: 14.5959 Joules
[Energy_Monitor] getConsumedJoules(0): 10.5076 Joules
[Energy_Monitor] getConsumedJoules(1): 0.000549316 Joules
[Energy_Monitor] getDRAMConsumedJoules: 0 Joules
[Energy_Monitor] getSystemConsumedJoules: 0 Joules
================== Energy Consumption by getALLCounterStates  ==================
[Energy_Monitor] getConsumedJoules: 14.5959 Joules
[Energy_Monitor] getConsumedJoules(0): 0 Joules
[Energy_Monitor] getConsumedJoules(1): 0 Joules
[Energy_Monitor] getDRAMConsumedJoules: 0 Joules
[Energy_Monitor] getSystemConsumedJoules: 7.3775 Joules
```
The system consumed joules is less that package consumed joules, what does the system consist of?
Thank you.

**Comment by rdementi** [comment]
Hi,

the value the system energy metric is retrieved from the MSR_PLATFORM_ENERGY_COUNTER register (Intel Software Developer Manual www.intel.com/sdm):
![image](https://github.com/user-attachments/assets/94e034f9-810c-4d02-a41b-4dfba9d79927)

Platform vendor and BIOS communicate the system energy value to the CPU MSR register. Could you please also share the output of pcm utility? Here is a sample output from an HP laptop (the platform and BIOS vendor is HP):

![image](https://github.com/user-attachments/assets/7179f7b6-7074-41ee-897d-153647391fcb)

The metrics of interest are "SYS energy",  "CPU energy", "PP0 energy" and "PP1 energy"

**Comment by rdementi** [comment]
I checked that the "SYS energy" PCM value on my HP laptop is close to the power meter reading at the wall.

**Comment by baihe-liu** [comment]
> I checked that the "SYS energy" PCM value on my HP laptop is close to the power meter reading at the wall.

Thank you for the information! It helps a lot. @rdementi 
Here is my output of pcm utility. I can get this 'SYS energy' by CLI, is there any way to get this metric  in code? I think this 'SYS energy' metric you showed me is different from the value in MSR_PLATFORM_ENERGY_COUNTER register, because I found the value in MSR times the energy unit is quite smaller than CPU energy.
![image](https://github.com/user-attachments/assets/0e7ef851-fbdf-4a5e-a5ff-be23dfefbbd4)

**Comment by rdementi** [comment]
The API used here is getSystemConsumedJoules :
https://github.com/intel/pcm/blob/aa70ef5f268c5bfd7e268cd711e6ee89c5a2318f/src/pcm.cpp#L342

The value from the MSR (MSR_SYS_ENERGY_STATUS = MSR_PLATFORM_ENERGY_COUNTER= 0x64D) is multiplied with the unit: 

https://github.com/intel/pcm/blob/aa70ef5f268c5bfd7e268cd711e6ee89c5a2318f/src/cpucounters.cpp#L1762

https://github.com/intel/pcm/blob/aa70ef5f268c5bfd7e268cd711e6ee89c5a2318f/src/cpucounters.cpp#L6982

https://github.com/intel/pcm/blob/aa70ef5f268c5bfd7e268cd711e6ee89c5a2318f/src/cpucounters.h#L3396

https://github.com/intel/pcm/blob/aa70ef5f268c5bfd7e268cd711e6ee89c5a2318f/src/cpucounters.h#L3487
